### PR TITLE
Rework Caching Guide

### LINKED ISSUE #652: Guides - Explain Hash Changes in Caching Guide
I got a scenario here, that I would like to discuss and hopefully create a source for docs, for what a developer should really think about to prevent causing hash changes.
I'm making something, that might seem like a very small change in my code, which has quite large effect on long-term caching.

I'm writing this issue in docs, because I'm sure it's expected, just that I want to be able to understand why this is happening, and hopefully help others as well.

Example repo is here (using `webpack@2.2.0-rc.7` and `webpack-chunk-hash@0.3.0`):
https://github.com/jouni-kantola/webpack-output-by-build-type/tree/why-hash-change

Here goes the scenario:
1) I build:

| Asset                                  | Size      | Chunks |           | 
|----------------------------------------|-----------|--------|-----------| 
| 0.9bbf83cf4b0bde5cdd21.dev.js.map      | 512 bytes | 0, 6   | [emitted] | 
| 0.9bbf83cf4b0bde5cdd21.dev.js          | 756 bytes | 0, 6   | [emitted] | 
| 2.a01031186175de379984.dev.js          | 455 bytes | 2, 6   | [emitted] | 
| 3.64a0d1a6257d07bf9c8e.dev.js          | 455 bytes | 3, 6   | [emitted] | 
| vendor.72eb1607924636054b21.dev.js     |  254 kB   | 4, 6   | [emitted] | 
| app.d67c37259cfc27adf14e.dev.js        | 16.9 kB   | 5, 6   | [emitted] | 
| 1.b255a8ecec8dd054c7f3.dev.js          | 756 bytes | 1, 6   | [emitted] | 
| 1.b255a8ecec8dd054c7f3.dev.js.map      | 512 bytes | 1, 6   | [emitted] | 
| 2.a01031186175de379984.dev.js.map      | 424 bytes | 2, 6   | [emitted] | 
| 3.64a0d1a6257d07bf9c8e.dev.js.map      | 430 bytes | 3, 6   | [emitted] | 
| vendor.72eb1607924636054b21.dev.js.map |  335 kB   | 4, 6   | [emitted] | 
| app.d67c37259cfc27adf14e.dev.js.map    | 18.9 kB   | 5, 6   | [emitted] | 
| index.html                             | 6.35 kB   |        | [emitted] | 

2) I change `module-a.js` from:
```
import is from 'is-thirteen';

export const a = () => console.log('module a says 12 + 1 === 13', is(12 + 1).thirteen());
```
to:
```
export const a = () => console.log('hello world');
```
3) I build yet again:

| Asset                                  | Size      | Chunks |           | 
|----------------------------------------|-----------|--------|-----------| 
| 0.3f19b2e7372b486647de.dev.js.map      | 512 bytes | 0, 6   | [emitted] | 
| 0.3f19b2e7372b486647de.dev.js          | 756 bytes | 0, 6   | [emitted] | 
| 2.a01031186175de379984.dev.js          | 455 bytes | 2, 6   | [emitted] | 
| 3.64a0d1a6257d07bf9c8e.dev.js          | 455 bytes | 3, 6   | [emitted] | 
| vendor.503164abc295ee5b7ca1.dev.js     | 254 kB    | 4, 6   | [emitted] | 
| app.928d94b1c06014c2234b.dev.js        | 16.9 kB   | 5, 6   | [emitted] | 
| 1.84b1d26493dcab2a97e9.dev.js          | 429 bytes | 1, 6   | [emitted] | 
| 1.84b1d26493dcab2a97e9.dev.js.map      | 370 bytes | 1, 6   | [emitted] | 
| 2.a01031186175de379984.dev.js.map      | 424 bytes | 2, 6   | [emitted] | 
| 3.64a0d1a6257d07bf9c8e.dev.js.map      | 430 bytes | 3, 6   | [emitted] | 
| vendor.503164abc295ee5b7ca1.dev.js.map | 335 kB    | 4, 6   | [emitted] | 
| app.928d94b1c06014c2234b.dev.js.map    | 18.9 kB   | 5, 6   | [emitted] | 
| index.html                             | 6.35 kB   |        | [emitted] | 

So to what I think needs to be documented:
1. What types of changes should a developer really be aware of when making changes.
2. In this build, is-thirteen changed ID from `59` to `85`, and that caused three bundles to be cache busted. In a larger app that could be quite a lot of code that needs to be shipped again to the end-user. Is there anyway to prevent this small change, from having this large effect?
3. Is there anyway of preventing the vendor bundle from changing hash, when changing consuming code?

I think I've done the most important bit, to extract the manifest, but still it's very easy to get chunks cache busted. There are loads of blogs discussing this, but I still believe the truly nitty gritty is missing to be documented.

### PR DESCRIPTION
**Pull request by skipjack**
Trying to combine the current [caching guide][1] with @timse's [post][2] and make it a bit easier to read with clear, fully-fleshed out examples. I'll be submitting a PR to @TheDutchCoder's [examples repo][3] soon as well for the small test project I created in relation to this.

Resolves #652 (hopefully -- and hopefully a bunch of the related open issues on the main repo)
Part of #1258 (will update that now)

[1]: https://webpack.js.org/guides/caching/
[2]: https://medium.com/webpack/predictable-long-term-caching-with-webpack-d3eee1d3fa31
[3]: https://github.com/TheDutchCoder/webpack-guides-code-examples

### CODE REVIEW

**Comment by skipjack** [comment]
@TheDutchCoder this should be ready for an initial review.

@timse I used your post in combination with the content that @okonet and @jouni-kantola already had here. Using the `HashedModuleIdsPlugin` allowed me to avoid (I think) certain steps in the post you wrote. I still need to test using `import()`, `externals` and a few other things, so another section or two may be warranted. It seems you have a lot of experience on the subject, so if you could help review that would be great.

@okonet @jouni-kantola I can't assign you, but if either of you can review that would be much appreciated.

Everyone should note that I also have a PR open at TheDutchCoder/webpack-guides-code-examples#17 that directly reflects the final version of the files/code in the guide. This should be a good testing ground for us to figure out exactly what does and doesn't mess up hashes. I want to keep things fairly straightforward so the guide doesn't go on forever, so maybe leaving some rare edge cases out is ok, but we should definitely try to cover all common use cases.

**Comment by skipjack** [comment]
I think once we get this to a good state, we should locate all related issues on the main repo (here I think it's only #652), see if they are fully or almost fully answered by the guide updates and close them with a note saying that people should read through the guide and, if any issues arise create a targeted issue with what needs to change.

As per @jouni-kantola's point in #652, it would be great if we could get to one source of truth for this documentation.

File: content/guides/caching.md [hunk 1]
```diff
@@ -22,7 +22,7 @@ related:
 
 T> This examples in this guide stem from [getting started](/guides/getting-started), [output management](/guides/output-management) and [code splitting](/guides/code-splitting).
```
Comment (TheDutchCoder) [inline content/guides/caching.md]: S/This examples/The examples

File: content/guides/caching.md [hunk 2]
```diff
@@ -22,7 +22,7 @@ related:
...
 
-So we're using webpack to bundle our modular application, deploying our `/dist` directory to the server, and clients, typically browsers, are hitting that server to grab the site and its assets. The last step can be time consuming, which is why browsers use a technique called [caching](). This allows sites to load faster with less unnecessary network traffic, however it can also cause headaches when you need new code to be picked up.
+So we're using webpack to bundle our modular application, deploying our `/dist` directory to the server, and clients, typically browsers, are hitting that server to grab the site and its assets. The last step can be time consuming, which is why browsers use a technique called [caching](https://en.wikipedia.org/wiki/Cache_(computing)). This allows sites to load faster with less unnecessary network traffic, however it can also cause headaches when you need new code to be picked up.
```
Comment (TheDutchCoder) [inline content/guides/caching.md]: Would rephrase this "...directory to the server, and clients, typically browsers, are hitting that server..." so it flows a little better.

File: content/guides/caching.md [hunk 3]
```diff
@@ -241,13 +241,13 @@ runtime.1400d5af64fc1b7b3a45.js    5.85 kB       2  [emitted]         runtime
     + 1 hidden module
 ```
 
-... we can see that all three have. This is because the [`module.id`]() of each is incremental by default, so...
+... we can see that all three have. This is because the [`module.id`](/api/module-variables#module-id-commonjs-) of each is incremental by default, so...
 
 - The `main` bundle changed because of it's new content.
```
Comment (TheDutchCoder) [inline content/guides/caching.md]: S/it's/its

File: content/guides/caching.md [hunk 4]
```diff
@@ -241,13 +241,13 @@ runtime.1400d5af64fc1b7b3a45.js    5.85 kB       2  [emitted]         runtime
...
 - The `vendor` bundle changed because it's `module.id` was changed.
```
Comment (TheDutchCoder) [inline content/guides/caching.md]: S/it's/its

File: content/guides/caching.md [hunk 5]
```diff
@@ -241,13 +241,13 @@ runtime.1400d5af64fc1b7b3a45.js    5.85 kB       2  [emitted]         runtime
...
 - And, the `runtime` bundle changed because it now contains a reference to a new module.
 
-The first and last are expected -- it's the `vendor` hash we want to fix. Luckily, there are two plugins that can help us out with this dilemma. First, the [`NamedModulesPlugin`]() which the path to the module rather than a numerical ID. While this plugin is useful during development for easier to read output, it does take a bit longer to run. The second option is the `HashedModuleIdsPlugin`, which is what we'll use as these examples are more targeted toward production builds:
+The first and last are expected -- it's the `vendor` hash we want to fix. Luckily, there are two plugins that can help us out with this dilemma. First, the [`NamedModulesPlugin`](/plugins/named-modules-plugin) which the path to the module rather than a numerical ID. While this plugin is useful during development for easier to read output, it does take a bit longer to run. The second option is the `HashedModuleIdsPlugin`, which is what we'll use as these examples are more targeted toward production builds:
```
Comment (TheDutchCoder) [inline content/guides/caching.md]: "...which the path..." seems like a word is missing here.

Comment (TheDutchCoder) [inline content/guides/caching.md]: Link to `HashedModuleIdsPlugin`?

File: content/guides/caching.md [hunk 6]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
-For an even more optimized setup:
+This guide focuses on the configuration changes needed to ensure that your `output` files are cached when appropriate, but re-requested when changed.
 
-1. Use compiler stats to get the file names when requiring resources in HTML.
-2. Generate the chunk manifest JSON and inline it into the HTML page before loading resources.
 
+## Output Filenames
 
-## The problem
+A simple way to ensure the browser picks up changed files is by using `output.filename` [substitutions](/configuration/output#output-filename). The `[hash]` substitution can be used to include a build-specific hash in the filename, however it's even better to use the `[chunkhash]` subsitution which include a bundle-specific hash in the filename.
```
Comment (TheDutchCoder) [inline content/guides/caching.md]: S/include/includes

File: content/guides/caching.md [hunk 7]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
++ import Print from './print';
 
-To fix that, we should use [`ChunkManifestWebpackPlugin`](https://github.com/diurnalist/chunk-manifest-webpack-plugin), which will extract the manifest to a separate JSON file. This replaces the chunk manifest with a variable in the webpack runtime. But we can do even better; we can extract the runtime into a separate entry by using `CommonsChunkPlugin`. Here is an updated `webpack.config.js` which will produce the manifest and runtime files in our build directory:
+  function component() {
+    var element = document.createElement('div');
 
-```js
-// webpack.config.js
-var ChunkManifestPlugin = require("chunk-manifest-webpack-plugin");
+    // Lodash, currently included via a script, is required for this line to work
```
Comment (TheDutchCoder) [inline content/guides/caching.md]: Remove this comment (I think)

File: content/guides/caching.md [hunk 8]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
+  import _ from 'lodash';
+- import Print from './print';
++ // import Print from './print';
 
-## Manifest inlining
+  function component() {
+    var element = document.createElement('div');
 
-Inlining the chunk manifest and webpack runtime (to prevent extra HTTP requests), depends on your server setup. There is a nice [walkthrough for Rails-based projects](https://brigade.engineering/setting-up-webpack-with-rails-c62aea149679). For server-side rendering in Node.js you can use [webpack-isomorphic-tools](https://github.com/halt-hammerzeit/webpack-isomorphic-tools).
+    // Lodash, currently included via a script, is required for this line to work
```
Comment (TheDutchCoder) [inline content/guides/caching.md]: Same here.

File: content/guides/caching.md [hunk 9]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
 
-## References
+## Conclusion
 
-* https://medium.com/@okonetchnikov/long-term-caching-of-static-assets-with-webpack-1ecb139adb95#.vtwnssps4
-* https://gist.github.com/sokra/ff1b0290282bfa2c037bdb6dcca1a7aa
-* https://github.com/webpack/webpack/issues/1315
-* https://github.com/webpack/webpack.js.org/issues/652
-* https://presentations.survivejs.com/advanced-webpack/
+Caching gets messy. Plain and simple. However the walk-through above should give you a running start to deploying consistent, cache-able assets. See the _Further Reading_ section below to learn more.
```
Comment (TheDutchCoder) [inline content/guides/caching.md]: S/cache-able/cachable

File: content/guides/caching.md [hunk 10]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
+  - title: Advanced Webpack Presentation
+    url: https://presentations.survivejs.com/advanced-webpack/
+  - title: Issue 1315
+    url: https://github.com/webpack/webpack/issues/1315
+  - title: Issue 652
+    url: https://github.com/webpack/webpack.js.org/issues/652
 ---
 
-To enable long-term caching of static resources produced by webpack:
+T> This examples in this guide stem from [getting started](/guides/getting-started), [output management](/guides/output-management) and [code splitting](/guides/code-splitting).
```
Comment (okonet) [inline content/guides/caching.md]: `This` and `examples` don't fit together grammatically

File: content/guides/caching.md [hunk 11]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
 
-1. Use `[chunkhash]` to add a content-dependent cache-buster to each file.
-2. Extract the webpack manifest into a separate file.
-3. Ensure that the entry point chunk containing the bootstrapping code doesn’t change hash over time for the same set of dependencies.
+So we're using webpack to bundle our modular application, deploying our `/dist` directory to the server, and clients, typically browsers, are hitting that server to grab the site and its assets. The last step can be time consuming, which is why browsers use a technique called [caching](https://en.wikipedia.org/wiki/Cache_(computing)). This allows sites to load faster with less unnecessary network traffic, however it can also cause headaches when you need new code to be picked up.
```
Comment (okonet) [inline content/guides/caching.md]: I'm not sure this guide needs this intro. It is good for a blog post but I'd expect this guide to get directly to the point. Probably replace it with a link to [my post](https://medium.com/@okonetchnikov/long-term-caching-of-static-assets-with-webpack-1ecb139adb95) and I could cross-link to this guide in the beginning to guide the traffic here?

File: content/guides/caching.md [hunk 12]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
 
-For an even more optimized setup:
+This guide focuses on the configuration changes needed to ensure that your `output` files are cached when appropriate, but re-requested when changed.
```
Comment (okonet) [inline content/guides/caching.md]: `This guide focuses on the configuration needed to ensure files produced by webpack compilation can remain cached unless their contents has changed.` ?

File: content/guides/caching.md [hunk 13]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
-But the problem here is, builds after *any file update* will update all filenames and clients will have to re-download all application code. So how can we guarantee that clients always get the latest versions of assets without re-downloading all of them?
+?> Add bash output
 
+Now our bundle's name should change to reflect its content, meaning if `src/index.js` is changed, the filename should change. However, if we run another build, we'll see that this is not the case:
 
-## Generating unique hashes for each file
+?> Add bash output
 
-What if we could produce the same filename, if the contents of the file did not change between builds? For example, it would be unnecessary to re-download a vendor file, when no dependencies have been updated, only application code.
+This is because webpack includes certain boilerplate, specifically the runtime and manifest, in the entry chunk.
```
Comment (okonet) [inline content/guides/caching.md]: In the previous sentence you're saying the file contents should have been changed but I think you meant to say "should not change"?

File: content/guides/caching.md [hunk 14]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
-module.exports = {
-  /*...*/
-  output: {
-    /*...*/
--   filename: "[name].[hash].js"
-+   filename: "[name].[chunkhash].js"
-  }
-};
-```
+## Extracting Boilerplate
```
Comment (okonet) [inline content/guides/caching.md]: `Extracting webpack boilerplate`?

File: content/guides/caching.md [hunk 15]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
+    ],
+    output: {
+      filename: '[name].[contenthash].js',
+      path: path.resolve(__dirname, 'dist')
+    }
+  };
 ```
 
-Although, each time we build for production, we’ll get different file names. Something, that looks like this:
+Now let's run a build to see if it worked:
```
Comment (okonet) [inline content/guides/caching.md]: By reading this sentence it's not clear if it worked or not and you're not clarifying it after the code block. I'd rephrase by making it less ambiguous.

File: content/guides/caching.md [hunk 16]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
-module.exports = {
-  /*...*/
-  plugins: [
-    function() {
-      this.plugin("done", function(stats) {
-        require("fs").writeFileSync(
-          path.join(__dirname, "build", "stats.json"),
-          JSON.stringify(stats.toJson()));
-      });
+Another thing we may want to extract is our core third-party libraries, in our case `lodash`, as they are less likely to change than our source code. This step will allow clients to request even less from the server to stay up to date. Let's do this using a combination of a new `entry` point along with another `CommonsChunkPlugin` instance:
```
Comment (okonet) [inline content/guides/caching.md]: Nitpicking but `thing` and `libraries` don't match to me. What about

> It's a good practice to extract third-party libraries, such as `lodash` or `react`, to a separate `vendor` chunk as they are less likely to change than our source code.

Comment (okonet) [inline content/guides/caching.md]: I'd replace `Let's do this ` with `This can be done by`

File: content/guides/caching.md [hunk 17]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
+      filename: '[name].[contenthash].js',
+      path: path.resolve(__dirname, 'dist')
     }
-  ]
-};
+  };
 ```
 
-Alternatively, just use one of these plugins to export JSON files:
+W> Note that order matters here. The `'vendor'` instance must be included prior to the `'runtime'` instance.
```
Comment (okonet) [inline content/guides/caching.md]: `instance`? Did you mean `chunk`?

File: content/guides/caching.md [hunk 18]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
-    new webpack.HashedModuleIdsPlugin(),
-    new WebpackChunkHash(),
-    new ChunkManifestPlugin({
-      filename: "chunk-manifest.json",
-      manifestVariable: "webpackManifest",
-      inlineManifest: true
-    })
-  ]
-};
+... we can see that all three have. This is because the [`module.id`](/api/module-variables#module-id-commonjs-) of each is incremental by default, so...
```
Comment (okonet) [inline content/guides/caching.md]: `is incremental by default` => `is not consistent by default`?

File: content/guides/caching.md [hunk 19]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
+
+- The `main` bundle changed because of it's new content.
```
Comment (okonet) [inline content/guides/caching.md]: Since this list is a part of the sentence, I'd make `The` lowercase and remove `.` in the end of each but last line

File: content/guides/caching.md [hunk 20]
```diff
@@ -4,284 +4,358 @@ sort: 11
...
+- The `vendor` bundle changed because it's `module.id` was changed.
+- And, the `runtime` bundle changed because it now contains a reference to a new module.
+
+The first and last are expected -- it's the `vendor` hash we want to fix. Luckily, there are two plugins that can help us out with this dilemma. First, the [`NamedModulesPlugin`](/plugins/named-modules-plugin) which the path to the module rather than a numerical ID. While this plugin is useful during development for easier to read output, it does take a bit longer to run. The second option is the [`HashedModuleIdsPlugin`](/plugins/hashed-module-ids-plugin), which is what we'll use as these examples are more targeted toward production builds:
```
Comment (okonet) [inline content/guides/caching.md]: I'd remove `dilemma`

Comment (okonet) [inline content/guides/caching.md]: `which the path to the module` is missing the verb

Comment (okonet) [inline content/guides/caching.md]: `ID` => `identifier`?

Comment (okonet) [inline content/guides/caching.md]: `for easier to read output` => `for simpler output` or `for the output that is easier to read`?

File: content/guides/caching.md (hunk 2 above)
Comment (skipjack) [inline content/guides/caching.md]: 👍

File: content/guides/caching.md (hunk 5 above)
Comment (skipjack) [inline content/guides/caching.md]: 👍

File: content/guides/caching.md (hunk 8 above)
Comment (skipjack) [inline content/guides/caching.md]: 👍

File: content/guides/caching.md (hunk 9 above)
Comment (skipjack) [inline content/guides/caching.md]: Yeah, I guess it's ok to append `able` to words even if you it's not an actual variation of that word. Using the `-` prevents spell checkers from throwing an error but I guess it does read better without it.

File: content/guides/caching.md (hunk 11 above)
Comment (skipjack) [inline content/guides/caching.md]: Hmmm I see your point but I think it is helpful to give a little context before jumping into things. I did use the link so as not to explain caching in too much detail -- also updated per the last review to make it flow a little better. Maybe we can still tighten it up a bit more? @TheDutchCoder what do you think?

File: content/guides/caching.md (hunk 12 above)
Comment (skipjack) [inline content/guides/caching.md]: 👍

File: content/guides/caching.md (hunk 13 above)
Comment (skipjack) [inline content/guides/caching.md]: 👍

File: content/guides/caching.md (hunk 15 above)
Comment (skipjack) [inline content/guides/caching.md]: 👍

File: content/guides/caching.md (hunk 16 above)
Comment (skipjack) [inline content/guides/caching.md]: Makes sense on both counts -- will update.

File: content/guides/caching.md (hunk 17 above)
Comment (skipjack) [inline content/guides/caching.md]: I meant `CommonsChunkPlugin` instance but maybe I should specify that.

File: content/guides/caching.md (hunk 18 above)
Comment (skipjack) [inline content/guides/caching.md]: No I did mean "incremental" in the sense that modules ids are simply incremented up in the order they are resolved. However I do see how this could be a bit confusing... I'll rephrase.

File: content/guides/caching.md (hunk 19 above)
Comment (skipjack) [inline content/guides/caching.md]: I see what you mean but the way I updated the sentence above makes this change unnecessary (I think).

File: content/guides/caching.md (hunk 14 above)
Comment (skipjack) [inline content/guides/caching.md]: Idk I think I'd rather leave this title as is for now. Partly because the mix of lowercase and uppercase (I know this is defined in our style guide/media repo) feels a bit odd but also I think it's pretty clear based on context.

**Comment by skipjack** [comment]
@okonet @TheDutchCoder awesome feedback, thank you both! Just pushed the updates...

What did you think of the article in general? My feeling is that it's a solid base that's now in line with our guides, but we may have to add a few of more of the oddities that can break caching, like those discussed in the various other guides we're linking to.

File: content/guides/caching.md (hunk 14 above)
Comment (jouni-kantola) [inline content/guides/caching.md]: Personally I'd prefer `Extract webpack runtime`. I interpret `boilerplate` like something that isn't variable. Also, later in the PR the asset is named `runtime`.

File: content/guides/caching.md [hunk 21]
```diff
@@ -4,284 +4,356 @@ sort: 11
...
-Hash: cfba4af36e2b11ef15db
-Version: webpack 2.2.0
-Time: 66ms
-                         Asset     Size  Chunks             Chunk Names
-vendor.50cfb8f89ce2262e5325.js  2.58 kB       0  [emitted]  vendor
-  main.70b594fe8b07bcedaa98.js  2.57 kB       1  [emitted]  main
-   [0] ./src/index.js 63 bytes {1} [built]
-   [1] ./src/vendor.js 63 bytes {0} [built]
-```
+As we learned in [code splitting](/guides/code-splitting), the [`CommonsChunkPlugin`](/plugins/commons-chunk-plugin) can be used to split modules out into separate bundles. A lesser-known feature of the `CommonsChunkPlugin` is extracting webpack's boilerplate and manifest which can change with every build. By specifying a name not mentioned in the `entry` configuration, the plugin will automatically extract what we want into a separate bundle:
```
Comment (jouni-kantola) [inline content/guides/caching.md]: Isn't the runtime added to the last specified commons chunk? Maybe that is only when specified as array.

File: content/guides/caching.md [hunk 22]
```diff
@@ -4,284 +4,356 @@ sort: 11
...
++       'lodash'
++     ]
++   },
+    plugins: [
+      new CleanWebpackPlugin(['dist']),
+      new HtmlWebpackPlugin({
+        title: 'Caching'
+      }),
++     new webpack.optimize.CommonsChunkPlugin({
++       name: 'vendor'
```
Comment (jouni-kantola) [inline content/guides/caching.md]: Don't we need `minChunks: Infinity` as property, or is it default?

File: content/guides/caching.md [hunk 23]
```diff
@@ -4,284 +4,356 @@ sort: 11
...
+``` diff
+  const path = require('path');
+  const CleanWebpackPlugin = require('clean-webpack-plugin');
+  const HtmlWebpackPlugin = require('html-webpack-plugin');
+
+  module.exports = {
+    entry: './src/index.js',
+    plugins: [
+      new CleanWebpackPlugin(['dist']),
+      new HtmlWebpackPlugin({
```
Comment (jouni-kantola) [inline content/guides/caching.md]: Maybe we should leave out `HtmlWebpackPlugin` from the caching docs?

**Comment by jouni-kantola** [comment]
One thing I'd like to point out is that this PR mostly describes how to setup a proper config, not what to be aware of during development, I.e. adding extra entries. What I've always found hardest is to keep hashes intact during development and refactoring. What seem like small changes can have bigger impact than expected (https://gist.github.com/jouni-kantola/1c1e2bfaebf30de50d1b6a71b869da13).

**Comment by timse** [comment]
Heya :) not sure about `HashedModuleIdsPlugin` might be working some magic that makes things I used unnecessary. I personally like the `NamedModulesPlugin`/`NamedModulesChunks` better for a couple of reasons, that dont necessarily count as a general guideline :)

File: content/guides/caching.md (hunk 23 above)
Comment (skipjack) [inline content/guides/caching.md]: I actually think the `HtmlWebpackPlugin` is key for this doc. If not, users have to track bundle names themselves using the manifest and I think that complicates the guide significantly. Plus, it's usage is already covered in `output-management` which comes before this guide.

File: content/guides/caching.md (hunk 14 above)
Comment (skipjack) [inline content/guides/caching.md]: How about `Extract Runtime & Libraries`? I think that pretty much covers the whole section. Could also be `Extract Runtime, Manifest, and Libraries` but the runtime and manifest are being extracted into the same bundle so I think it's ok to just use `Runtime`.

File: content/guides/caching.md (hunk 21 above)
Comment (skipjack) [inline content/guides/caching.md]: It's extracted to a separate chunk when you use a name that doesn't already exist (i.e. no entry chunk exists by that name). I played around with the array syntax as well but ran into some issues trying to do it all with one instance.

File: content/guides/caching.md (hunk 22 above)
Comment (skipjack) [inline content/guides/caching.md]: See the example at TheDutchCoder/webpack-guides-code-examples#17... I don't think it's necessary but I also didn't test chunks too much yet (though I do have a bit more code to push that uses `import()` to generate a child chunk).

**Comment by skipjack** [comment]
> One thing I'd like to point out is that this PR mostly describes how to setup a proper config, not what to be aware of during development, I.e. adding extra entries. What I've always found hardest is to keep hashes intact during development and refactoring. What seem like small changes can have bigger impact than expected.

@jouni-kantola I see your point, but I mean ideally we want to provide the configuration changes needed so that unexpected hash changes don't occur, right? Maybe some discussion in the conclusion of what to watch out for would help though?

> Heya :) not sure about `HashedModuleIdsPlugin` might be working some magic that makes things I used unnecessary. I personally like the `NamedModulesPlugin` / `NamedModulesChunks` better for a couple of reasons, that dont necessarily count as a general guideline :)

@timse I used the `HashedModuleIdsPlugin` because I found it simpler when testing (TheDutchCoder/webpack-guides-code-examples#17) and it seemed I didn't need to include any of the `Named***` plugins. That said, I did see what you said in your post about the `Named***` plugins yielding slightly smaller bundles.

> That being said, I would use the named chunks plugin - But probably with a saner fallback function than specified in the guide a wrote :)

I will take a look at that plugin tomorrow. I'm going to merge this for now but I definitely would like to keep the conversation going and do some follow up PRs. Maybe we can all test the example I put together in TheDutchCoder/webpack-guides-code-examples#17 and see if we can poke holes any holes in it? I don't think we have to cover extreme edge cases but if there are some common scenarios that break it we should definitely address them in the guide.
